### PR TITLE
perf(torch): dedupe double trainable_weights walk in train_step

### PR DESCRIPTION
**Pull request by pctablet505**
perf(torch): dedupe double trainable_weights walk in train_step

## Why this was closed

This change showed no measurable macro benefit on either axis. Its isolated
eager effect is within measurement noise on the models tested (RTX 4050,
torch 2.13.0+cu130, CNN forward / LLM forward / LLM generate-32), and it is
compile-neutral: 0 graph-break delta versus master, with a deterministic
dynamo decomposition (see the per-PR compile scorecard). It was backed only
by a deterministic micro-benchmark showing a call-count reduction (3 to 2
`trainable_weights` walks per `train_step`), not by an observed wall-clock
or compile-graph effect.

torch.compile graph breaks across the #22561 series (this PR does not change the count)

Relates to #22561. Sits in the same area as #23187's loss/trainer fast
path but touches a disjoint hunk of `train_step`, so it's not folded into
that PR.

## Summary

`TorchTrainer.train_step` read `self.trainable_weights` twice back to
back — once for the `if self.trainable_weights:` guard, once more for
the `[:]` snapshot handed to the optimizer. It's a full recursive
layer-tree walk, not a cached property, so this did the walk twice for
no reason. Now the guard site caches the result in a local and both
sites reuse it.

## Why it's safe

Nothing mutates the layer tree between the two former read sites within
a single `train_step` call, so a cached read is equivalent to two
independent reads (verified with an identity-equal check across
back-to-back reads). Change is scoped to `keras/src/backend/torch/trainer.py`
only — no other backend or file touched — and is line-disjoint from
#23187's hunks in the same file (clean `git merge-tree` against its
branch).

## Benchmarks

PR #23299: master vs PR alone vs PR+parents, Keras torch GPU eager, ms

Deterministic proof: property-walk count per `train_step` call drops
3 → 2. Estimated ~0.6% of step wall time — smallest-effect candidate of
this round's seven; GPU wall-clock deltas (RTX 4050, torch 2.13.0+cu130,
8-pair interleaved A/B) land in noise for all three workloads (CNN
forward, LLM forward, LLM generate-32), as expected at that magnitude.
Forward outputs and generate-32 token ids are bit-identical to master.

## Tests

Added torch-only tests in `keras/src/backend/torch/trainer_test.py`: a
property-access counting test (asserts exactly 2 reads per step, down
from 3), an identity-equality check across back-to-back reads, and an
end-to-end `train_on_batch` reproducibility check under a fixed seed.

## Contributor Agreement

- [x] I am a human, and not a bot.
- [x] I will be responsible for responding to review comments in a timely manner.
- [x] I will work with the maintainers to push this PR forward until submission.

### CODE REVIEW

File: keras/src/backend/torch/trainer.py [hunk 1]
```diff
@@ -64,12 +64,13 @@ def train_step(self, data):
...
-        if self.trainable_weights:
+        trainable_weights = self.trainable_weights
+        if trainable_weights:
             # Call torch.Tensor.backward() on the loss to compute gradients
             # for the weights.
             loss.backward()
 
-            trainable_weights = self.trainable_weights[:]
+            trainable_weights = trainable_weights[:]
             gradients = [v.value.grad for v in trainable_weights]
```
[the remark this answers is not in the document]
Comment (pctablet505) [reply keras/src/backend/torch/trainer.py]: Good catch -- `trainable_weights` is already a freshly-built list by that point (`Layer.trainable_weights` is a list comprehension, not cached), so the extra `[:]` was copying a list nothing else references. Done in 6db53ac -- removed the redundant `[:]` and use `trainable_weights` directly in the gradients comprehension.

**Comment by pctablet505** [comment]
Closing this one. It's a trainer-only change off the forward/inference path, within measurement noise, backed only by a call-count reduction. Trimming the #22561 series to the changes with demonstrable macro impact so maintainer review stays focused on those.
